### PR TITLE
[deepspeed] performance docs

### PR DESCRIPTION
**Pull request by stas00**
This PR extends the deepspeed docs to include:

1. performance docs
2. network communication docs

### CODE REVIEW

File: docs/source/en/main_classes/deepspeed.mdx [hunk 1]
```diff
@@ -795,6 +795,39 @@ default value in the following cases:
    the increased data buffers.
 
 
+#### ZeRO-0 Config
+
+I'm listing Stage 0 and 1 last since they are rarely used.
```
Comment (sgugger) [inline docs/source/en/main_classes/deepspeed.mdx:800]: I'm not sure what this brings to the doc. Let's remove this sentence?

File: docs/source/en/main_classes/deepspeed.mdx [hunk 2]
```diff
@@ -1117,6 +1150,53 @@ values look like, but we highly recommend using the one with multiple `auto` set
 }
 ```
 
+#### How to Choose Which ZeRO Stage and Offloads To Use For Best Performance
+
+So now you know there are all these different stages and how do I know which of them to use. This section will attempt to address this question.
```
Comment (sgugger) [inline docs/source/en/main_classes/deepspeed.mdx:1155]:
```suggestion
So now you know there are all these different stages. How to decide which of them to use? This section will attempt to address this question.
```

File: docs/source/en/main_classes/deepspeed.mdx [hunk 3]
```diff
@@ -1117,6 +1150,53 @@ values look like, but we highly recommend using the one with multiple `auto` set
...
+
+8. Definitely use mixed half-precision over fp32 - so bf16 on Ampere and higher GPUs and fp16 on older gpu architectures.
+
+9. If you still OOM you could add more hardware or enable ZeRO-Infinity - that is switch offloads `offload_param` and  `offload_optimizer` to `nvme`. You need to make sure it's a very fast nvme. As an anecdote I was able to infer BLOOM-176B on a tiny GPU using ZeRO-Infinity except it was extremely slow. But it worked!
+
+You can, of course, work through these steps in reverse by starting with the most GPU memory efficient config and then going backwards. Or try bi-secting it.
+
+Once you have your BS=1 not leading to OOM, measure your effective throughput.
+
+Next try to increase the BS to as large as you can, since the higher the batch size the more efficient the GPUs are as they perform the best when matrices they multiply are huge.
```
Comment (sgugger) [inline docs/source/en/main_classes/deepspeed.mdx:1188]: We're in the doc, let's spell batch size since there is no character limit :-)

File: docs/source/en/main_classes/deepspeed.mdx [hunk 4]
```diff
@@ -1117,6 +1150,53 @@ values look like, but we highly recommend using the one with multiple `auto` set
...
+
+Now the performance optimization game starts. You can turn off some offload features or step down in ZeRO stages and increase/decrease BS and again measure your effective throughput. Rinse and repeat until satisfied.
+
+Don't spend forever on it, but if you're about to start a 3 months training - do spend a few days on it to find the most effective throughput-wise setup. So that your training cost will be the lowest and you will finish training faster. In the current crazy-paced ML world, if it takes you an extra month to train something you are likely to miss a golden opportunity. Of course, this is only me sharing an observation and in no way I'm trying to rush you. Before beginning to train BLOOM-176B I spent 2 days on this process and was able to increase throughput from 90 to 150 TFLOPs! This effort saved us more than one month of training time.
+
+These notes were written primarily for the training mode, but they should mostly apply for inference as well. For example, during inference Gradient Checkpointing is a no-op since it is only useful during training.
```
Comment (sgugger) [inline docs/source/en/main_classes/deepspeed.mdx:1194]: If not training or not using DeepSpeed inference, we should recommend Accelerate, as it was faster in the tests.

**Review by sgugger — COMMENTED** [review]
Thanks for your PR! I have a few comments.

File: docs/source/en/main_classes/deepspeed.mdx (hunk 4 above)
Comment (stas00) [reply docs/source/en/main_classes/deepspeed.mdx:1194]: does this work?

```suggestion
These notes were written primarily for the training mode, but they should mostly apply for inference as well. For example, during inference Gradient Checkpointing is a no-op since it is only useful during training. Additionally, we found out that if you 
are doing a multi-GPU inference and not using [DeepSpeed-Inference](https://www.deepspeed.ai/tutorials/inference-tutorial/), [Accelerate](https://huggingface.co/blog/bloom-inference-pytorch-scripts) should provide a superior performance.

```

File: docs/source/en/main_classes/deepspeed.mdx (hunk 1 above)
Comment (stas00) [reply docs/source/en/main_classes/deepspeed.mdx]: Perfect, thank you, Sylvain.

**Review by sgugger — APPROVED** [review]
